File: elixir/reverse_odd_levels.exs
# https://leetcode.com/problems/reverse-odd-levels-of-binary-tree/
# 2024/12/20

Code.require_file("tree_node.ex", __DIR__)

defmodule Solution do
  @spec reverse_odd_levels(root :: TreeNode.t() | nil) :: TreeNode.t() | nil
  def reverse_odd_levels(nil), do: nil

  def reverse_odd_levels(root) do
    if root.left == nil or root.right == nil do
      root
    else
      {new_left, new_right} = dfs(root.left, root.right, 1)
      %TreeNode{root | left: new_left, right: new_right}
    end
  end

  defp dfs(nil, nil, _level), do: {nil, nil}
  defp dfs(nil, right, _level), do: {nil, right}
  defp dfs(left, nil, _level), do: {left, nil}

  defp dfs(left, right, level) do
    lval = left.val
    ll = left.left
    lr = left.right

    rval = right.val
    rl = right.left
    rr = right.right

    {new_left_val, new_right_val} =
      if rem(level, 2) == 1 do
        {rval, lval}
      else
        {lval, rval}
      end

    {updated_ll, updated_rr} = dfs(ll, rr, level + 1)
    {updated_lr, updated_rl} = dfs(lr, rl, level + 1)

    new_left = %TreeNode{val: new_left_val, left: updated_ll, right: updated_lr}
    new_right = %TreeNode{val: new_right_val, left: updated_rl, right: updated_rr}

    {new_left, new_right}
  end
end

defmodule Main do
  def main do
    test_cases = [
      %{
        root: [2, 3, 5, 8, 13, 21, 34],
        expected: [2, 5, 3, 8, 13, 21, 34]
      },
      %{
        root: [7, 13, 11],
        expected: [7, 11, 13]
      },
      %{
        root: [0, 1, 2, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2],
        expected: [0, 2, 1, 0, 0, 0, 0, 2, 2, 2, 2, 1, 1, 1, 1]
      }
    ]

    test_cases
    |> Enum.each(fn %{root: root, expected: expected} ->
      result =
        root
        |> TreeNodeHelpers.from_list()
        |> Solution.reverse_odd_levels()
        |> TreeNodeHelpers.to_list()

      IO.puts("Input: root = #{inspect(root)}")
      IO.puts("Output: #{inspect(result)}")
      IO.puts("Expected: #{inspect(expected)}")
      IO.puts("Test #{if result == expected, do: "Passed", else: "Failed"}")
      IO.puts("------")
    end)
  end
end

Main.main()

File: rust/src/bin/reverse_odd_levels.rs
// https://leetcode.com/problems/reverse-odd-levels-of-binary-tree/
// 2024/12/20

mod tree_node;
use std::cell::RefCell;
use std::rc::Rc;
use tree_node::TreeNode;
impl Solution {
    pub fn reverse_odd_levels(
        root: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        if let Some(root_rc) = root.clone() {
            let root_ref = root_rc.borrow();
            Self::dfs(&root_ref.left, &root_ref.right, 1);
        }

        root
    }

    fn dfs(
        left_node: &Option<Rc<RefCell<TreeNode>>>,
        right_node: &Option<Rc<RefCell<TreeNode>>>,
        level: i32,
    ) {
        match (left_node, right_node) {
            (None, None) => return,
            (Some(left), Some(right)) => {
                if level % 2 == 1 {
                    let (mut left_borrow, mut right_borrow) =
                        (left.borrow_mut(), right.borrow_mut());
                    std::mem::swap(&mut left_borrow.val, &mut right_borrow.val);
                }

                Self::dfs(&left.borrow().left, &right.borrow().right, level + 1);
                Self::dfs(&left.borrow().right, &right.borrow().left, level + 1);
            }
            _ => panic!("Invalid input! Need perfect tree"),
        }
    }
}

struct Solution {}

fn main() {
    println!(
        "{:?}",
        TreeNode::to_array(Solution::reverse_odd_levels(TreeNode::from_array(
            &[2, 3, 5, 8, 13, 21, 34].map(|e| Some(e))
        )))
    ); // [2,5,3,8,13,21,34]

    println!(
        "{:?}",
        TreeNode::to_array(Solution::reverse_odd_levels(TreeNode::from_array(
            &[7, 13, 11].map(|e| Some(e))
        )))
    ); // [7,11,13]

    println!(
        "{:?}",
        TreeNode::to_array(Solution::reverse_odd_levels(TreeNode::from_array(
            &[0, 1, 2, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2].map(|e| Some(e))
        )))
    ); // [0,2,1,0,0,0,0,2,2,2,2,1,1,1,1]
}
